kate/grid/src/dims.rs

```
@@ -0,0 +1,104 @@
+		Self {
+			width: nonzero_unchecked(width),
+			height: nonzero_unchecked(height),
+		}
+	}
+
+	pub fn width(&self) -> usize {
+		self.width.get()
+	}
+
+	pub fn width_nz(&self) -> NonZeroUsize {
+		self.width
+	}
+
+	pub fn height(&self) -> usize {
+		self.height.get()
+	}
+
+	pub fn height_nz(&self) -> NonZeroUsize {
+		self.height
+	}
+
+	pub fn n_cells(&self) -> usize {
+		self.width.saturating_mul(self.height).get()
+	}
+
+	pub fn divides(&self, other: &Self) -> bool {
+		other.width.get() % self.width == 0 && other.height.get() % self.height == 0
+	}
+
+	pub fn extend(&self, e: Extension) -> Self {
+		Self {
+			width: e.width_factor.saturating_mul(self.width),
+			height: e.height_factor.saturating_mul(self.height),
+		}
+	}
+}
+
+/// The ways a set of dimensions can be extended
+#[derive(Debug, Clone)]
+pub struct Extension {
+	/// This means extending the height of the grid by some factor.
+	/// `2` would mean doubling the grid upwards, increasing the height by a factor of
+	/// 2 and multiplying the number of rows by 2
+	pub height_factor: NonZeroUsize,
+	/// This means extending the width of the grid by some factor.
+	/// `2` would mean doubling the grid sideways, increasing the width by a factor of
+	/// 2 and multiplying the number of columns by 2
+	pub width_factor: NonZeroUsize,
+}
+
+impl Extension {
+	pub const fn height(factor: NonZeroUsize) -> Self {
+		Self {
+			height_factor: factor,
+			width_factor: nonzero_unchecked(1),
+		}
+	}
+
+	pub const fn height_unchecked(factor: usize) -> Self {
+		Self {
+			height_factor: nonzero_unchecked(factor),
+			width_factor: nonzero_unchecked(1),
+		}
+	}
+
+	pub const fn width(factor: NonZeroUsize) -> Self {
+		Self {
+			height_factor: nonzero_unchecked(1),
+			width_factor: factor,
+		}
+	}
+
+	pub const fn width_unchecked(factor: usize) -> Self {
+		Self {
+			height_factor: nonzero_unchecked(1),
+			width_factor: nonzero_unchecked(factor),
```

Same as `Dimensions::new_unchecked`, it could panic. You could use `unsafe` if you want.

---

I would rather have it panic then violate memory rules. The reason `NonZeroUsize::new_unchecked` is unsafe is that it can do optimizations: for example `Option<NonZeroUsize>` is just a `usize` where 0 is `None`. Here we should panic if 0 is given, not fail silently while potentially corrupting memory.